Quarantined Submission Tables

mauberti-bc opened the pull request:
## Jira Tickets

- [SIMSBIOHUB-795](https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-795)

## Description of Changes

Adds new tables for tracking submission uploads.

**NOTE**: ```upload-ingestion-service.ts``` shows the business logic and how all the tables are used.

**New Tables / Enums:**

- **Enums**
  - `upload_status` – lifecycle of an upload session (`pending`, `completed`, `aborted`, `expired`)
  - `process_status` – general processing/lifecycle status used for records that undergo background proceesing (`draft`, `pending`, `completed`, `failed`)
  - `security_status` – security/quarantine result status for artifacts and per-file scans (`pending`, `clean`, `infected`, `error`, `skipped`)
  - `upload_artifact_role` – role of a file in the archive (feature or attachment)

- **Tables**
  - `upload` – stores upload sessions, with status, S3 ID, timestamps, and user info.
  - `artifact` – record of uploaded objects in storage, with checksum, byte size, and processing status.
  - `upload_archive` – links a specific archive artifact to an upload session.
  - `upload_artifact` – join table linking uploads to their produced artifacts, excluding archive files (eg. tarball)
  - `artifact_security` – tracks the current security status of an artifact. One active record per artifact.
  - `artifact_security_scan` – tracks individual malware scan events for artifacts.
  - `artifact_security_scan_file` – stores per-file malware scan results for scanned artifacts.
  - `submission_upload` – links submission records to the `upload_id` that created it.

**Endpoint Changes / Usage:**

- **POST `/submission/upload/archive`** – Creates a new upload session and submission ID.
- **PUT `/submission/upload/{uploadId}`** – Marks upload as complete. Triggers downstream processing for parsing the tarball and malware scanning.

**Testing / Notes:**

- The admin page at [http://localhost:7200/admin/submissions/create](http://localhost:7200/admin/submissions/create) works for creating an archive submission.

dylanrogowsky-oxd commented on api/src/services/quarantine/quarantine-service.ts:
```diff
@@ -0,0 +1,269 @@
+    });
+
+    // 7. Copy the submission from the quarantined bucket to the main bucket
+    await this.objectStorageService.promoteFromQuarantine(record.uri);
```
Does this always promote the file, even if it's infected

dylanrogowsky-oxd commented on api/src/paths/submission/quarantine/index.ts:
Should this be `quarantineId`?
```
export interface PresignedUploadUrlResponse {
  quarantineId: string;
  uploadId: string;
  key: string;
  partSizeBytes: number;
  partCount: number;
  presignedUrls: PresignedUrl[];
}
```

mauberti-bc committed:
rebase

fix tests

code smells

remove duplicate type

fix bad import

fix another bad import

mauberti-bc committed:
revert some changes

dylanrogowsky-oxd commented on api/src/repositories/upload/upload-artifact-repository.ts:
```diff
@@ -0,0 +1,152 @@
+        upload_id = COALESCE(${uploadArtifact.upload_id}, upload_id),
+        artifact_id = COALESCE(${uploadArtifact.artifact_id}, artifact_id),
+        role = COALESCE(${uploadArtifact.role}, role),
+        upload_archive_id = COALESCE(${uploadArtifact.upload_archive_id}, upload_archive_id),
```
I think the trailing comma should be removed

mauberti-bc replied:
Fixed

dylanrogowsky-oxd commented on api/src/models/upload-artifact.ts:
```diff
@@ -0,0 +1,43 @@
+
+export const UploadArtifactRoleZod = z.enum(['feature', 'attachment']);
+
+export enum UploadArtifactRoleEnum {
```
Is `archive` missing?
```
    -- Type of upload artifact
    CREATE TYPE upload_artifact_role AS ENUM (
      'feature',  
      'attachment',
      'archive'
    );
    ```

dylanrogowsky-oxd commented on api/src/services/upload/upload-ingestion-service.ts:
```diff
@@ -0,0 +1,177 @@
+      artifact_id: artifact.artifact_id,
+      status: ProcessStatusStatusEnum.DRAFT
+    });
+
```
Should it create an `upload_artifact` record after Step 5 to link the upload and the artifact?

dylanrogowsky-oxd commented on api/src/models/artifact-quarantine.ts:
```diff
@@ -0,0 +1,43 @@
+  artifact_quarantine_id: z.string().uuid(),
+  upload_artifact_id: z.string().uuid(),
+  status: SecurityResultStatusZod
+});
```
Is there a disconnect with the table? Eg `upload_artifact_id` vs `artifact_id`?

```
    CREATE TABLE artifact_quarantine (
      artifact_quarantine_id uuid DEFAULT public.gen_random_uuid(),
      artifact_id uuid NOT NULL,
      status security_status,
      create_date timestamptz(6) DEFAULT now() NOT NULL,
      create_user integer NOT NULL,
      update_date timestamptz(6),
      update_user integer,
      revision_count integer DEFAULT 0 NOT NULL,
      CONSTRAINT artifact_quarantine_pk PRIMARY KEY (artifact_quarantine_id),
      CONSTRAINT artifact_quarantine_artifact_fk FOREIGN KEY (artifact_id) REFERENCES artifact(artifact_id),
      CONSTRAINT artifact_quarantine_artifact_uq UNIQUE (artifact_id)
    );
    ```

mauberti-bc replied:
I intentionally designed the quarantine / malware-scanning flow to depend only on ```artifact``` records. To make this clearer, `artifact_quarantine.status` could be renamed to `artifact_quarantine.security_status`.

### Lifecycle flow

- When a tarball artifact is first created:
  - `artifact.status = 'draft'`
- Once the upload is verified:
  - `artifact.status = 'pending'`
  - `uploaded_at = now()`

At that point, a corresponding row is inserted into `artifact_quarantine`:

- `artifact_id = <artifact.artifact_id>`
- `status = 'pending'` (awaiting malware scan)
- This ```artifact.status``` represents whether the file has been uploaded. For clarity, we could give it its own enum (or check constraint) with these options: `draft`, `uploaded`, `deleted`.

`artifact_quarantine` represents the current security state of the artifact. I was thinking that all malware scanning logic should reference `artifact_id`, not `upload_artifact_id`, since security applies to the stored object itself, not the upload session that produced it.

Individual scan executions and detailed results are tracked in `artifact_quarantine_scan`. When a scan completes, `artifact_quarantine.status` is updated to one of: `clean`, `infected`, `error`, or `skipped`, in addition to the `artifact_quarantine_scan.status` being one of: `draft`, `pending`, `completed`, `failed`.

We could enforce only one active security record per artifact:

```sql
CREATE UNIQUE INDEX artifact_quarantine_uk
  ON artifact_quarantine (artifact_id)
  WHERE record_end_date IS NULL;

dylanrogowsky-oxd commented on api/src/repositories/upload/upload-repository.ts:
```diff
@@ -0,0 +1,135 @@
+   * Insert a new upload
+   */
+  async insertUpload(upload: CreateUpload): Promise<{ upload_id: string }> {
+    const sqlStatement = SQL`
+      INSERT INTO upload (
+        status,
+        s3_upload_id,
+        record_end_date
+      ) VALUES (
+        ${upload.status},
+        ${upload.s3_upload_id},
+        ${upload.record_end_date ?? null}
+      )
+      RETURNING upload_id;
+    `;
```
Is there a potential issue with `record_end_date` possibly being null and no value for `create_user`?

```
    CREATE TABLE upload (
      upload_id uuid DEFAULT public.gen_random_uuid(),
      status upload_status NOT NULL,
      s3_upload_id VARCHAR(100),
      record_end_date timestamptz NOT NULL,
      create_date timestamptz(6) DEFAULT now() NOT NULL,
      create_user integer NOT NULL,
      update_date timestamptz(6),
      update_user integer,
      revision_count integer DEFAULT 0 NOT NULL,
      CONSTRAINT upload_pk PRIMARY KEY (upload_id)
    );
```

mauberti-bc replied:
```record_end_date``` is required because all presigned URLs should expire. Upload records are currently valid for 30 minutes.

The ```POST submission/upload/archive``` endpoint requires a user, which will be the SIMS service account typically. Even if the upload endpoint didn't require auth, ```create_user``` would then equal the ```system_user_id``` of the internal API.

dylanrogowsky-oxd commented on api/src/models/upload-artifact.ts:
```diff
@@ -0,0 +1,43 @@
+
+export const UploadArtifactRoleZod = z.enum(['feature', 'attachment']);
+
+export enum UploadArtifactRoleEnum {
+  FEATURE = 'feature',
+  ATTACHMENT = 'attachment'
+}
```
Is `archive` missing?
```
    -- Type of upload artifact
    CREATE TYPE upload_artifact_role AS ENUM (
      'feature',  
      'attachment',
      'archive'
    );
```

mauberti-bc replied:
I was flip-flopping on whether to track archive files in the ```upload_artifact``` table, and most recently landed on: we shouldn't. If the archive just facilitates uploading many files, we can just track it in ```upload_archive```. When the archive file is opened, we would insert each contained file as a record in ```upload_artifact```. Right now, the archive only has features and attachments that the features reference, so we only need the two.

dylanrogowsky-oxd commented on api/src/repositories/upload/upload-artifact-repository.ts:
```diff
@@ -0,0 +1,152 @@
+   * @returns {Promise<UploadArtifact>} - The upload artifact record.
+   * @throws {ApiExecuteSQLError} - Throws an error if the upload artifact is not found.
+   */
+  async getUploadArtifact(uploadArtifactId: string): Promise<UploadArtifact> {
+    const sqlStatement = SQL`
+      SELECT
+        upload_artifact_id,
+        upload_id,
+        artifact_id,
+        role
+      FROM
+        upload_artifact
+      WHERE
+        upload_artifact_id = ${uploadArtifactId};
+    `;
```
For completeness, should `upload_archive_id` also be in the SELECT?

```
    CREATE TABLE upload_artifact (
      upload_artifact_id uuid DEFAULT public.gen_random_uuid(),
      upload_id uuid NOT NULL,
      artifact_id uuid NOT NULL,
      role upload_artifact_role NOT NULL,
      upload_archive_id uuid,
      ...
    )
```

mauberti-bc replied:
Yes, thank you

dylanrogowsky-oxd commented on api/src/repositories/upload/upload-artifact-repository.ts:
```diff
@@ -0,0 +1,152 @@
+   *
+   * @returns {Promise<UploadArtifact[]>} - A list of all upload artifacts.
+   */
+  async getUploadArtifacts(): Promise<UploadArtifact[]> {
+    const sqlStatement = SQL`
+      SELECT
+        upload_artifact_id,
+        upload_id,
+        artifact_id,
+        role
+      FROM
+        upload_artifact;
+    `;
```
For completeness, should `upload_archive_id` also be in the SELECT?

```
    CREATE TABLE upload_artifact (
      upload_artifact_id uuid DEFAULT public.gen_random_uuid(),
      upload_id uuid NOT NULL,
      artifact_id uuid NOT NULL,
      role upload_artifact_role NOT NULL,
      upload_archive_id uuid,
      ...
    )
```

dylanrogowsky-oxd commented on api/src/services/upload/upload-ingestion-service.ts:
```diff
@@ -0,0 +1,177 @@
+      upload_id
+    });
+
+    // 4. Create placeholder artifact for archive
+    const key = `submissions/${submission_id}/uploads/${upload_id}.tar`;
+    const artifact = await this.artifactService.insertArtifact({
+      bucket: getQuarantineObjectStoreBucketName(),
+      status: ArtifactStatusEnum.PENDING,
+      object_key: key,
+      byte_size: bytes,
+      checksum_sha256: null,
+      uploaded_at: null
+    });
```
Should it create an `upload_artifact` record after Step 4 to link the upload and the artifact?

mauberti-bc replied:
I've been treating ```upload_artifact``` as the parsed files inside the archive, but I think your implied assumption that upload_artifact represents all the artifacts associated to an upload is better.

- ```upload_artifact ```contains all uploaded files, including archives
- ```upload_archive``` acts like the queue for archives that need to be opened, after which their contents will be inserted as new records in ```upload_artifact```

mauberti-bc committed:
clean up status types & PR comments

mauberti-bc committed:
replace quarantine -> security

mauberti-bc committed:
replace quarantine -> security in file paths

mauberti-bc committed:
fix sql

mauberti-bc committed:
cleanup status property on scan files

mauberti-bc committed:
fix status type

mauberti-bc committed:
fix numbering

mauberti-bc committed:
misc naming improvements

mauberti-bc commented:
Alright, should be good now:
- Renamed generic `status` columns to more specific ones (eg. `upload_status` and `archive_status`) to make the schema clearer and avoid confusion when multiple statuses are involved in the same query.
- Added a `BLOCKED` state to the processing lifecycle so background tasks can be queued and ready without actually running until their prerequisites are met.
- Cleaned up and fixed the archive upload completion flow: we now authorize the request, mark the upload and artifacts as completed, queue artifacts for malware scanning, block archive processing until scans finish, and then finalize the multipart upload in the security S3 bucket.
- Added a `FAILED` state to `upload_status`, in case some part of the upload provisioning (eg. generating presigned URLs) fails
- Added an `ARCHIVED` state to `artifact_status`, in case we move old files

dylanrogowsky-oxd commented on api/src/repositories/upload/artifact-security-scan-file-repository.ts:
```diff
@@ -0,0 +1,157 @@
+
+    sqlStatement.append(
+      SQL` RETURNING artifact_security_scan_file_id, artifact_security_scan_id, file_path, security;`
+    );
```
I think `security` should be `result`

```
CREATE TABLE artifact_security_scan_file (
  artifact_security_scan_file_id uuid DEFAULT public.gen_random_uuid(),
  artifact_security_scan_id uuid NOT NULL,
  file_path text NOT NULL,
  result security_status NOT NULL,
  ...
);
```

dylanrogowsky-oxd commented on api/src/repositories/upload/artifact-security-scan-file-repository.ts:
```diff
@@ -0,0 +1,157 @@
+    uploadArtifactSecurityScanFileId: string,
+    scanFile: UpdateArtifactSecurityScanFile
+  ): Promise<ArtifactSecurityScanFile> {
+    const sqlStatement = SQL`
+      UPDATE biohub.artifact_security_scan_file
+      SET
+        security = COALESCE(${scanFile.result}, security)
+      WHERE
+        artifact_security_scan_file_id = ${uploadArtifactSecurityScanFileId}
+      RETURNING 
+        artifact_security_scan_file_id,
+        artifact_security_scan_id,
+        file_path,
+        result;
+    `;
```
I think `security` on line 136 should be `result`

```
CREATE TABLE artifact_security_scan_file (
  artifact_security_scan_file_id uuid DEFAULT public.gen_random_uuid(),
  artifact_security_scan_id uuid NOT NULL,
  file_path text NOT NULL,
  result security_status NOT NULL,
  ...
);
```

mauberti-bc committed:
change security -> result

fix reference

dylanrogowsky-oxd reviewed (commented):
`api/src/repositories/upload/submission-upload-repository.ts`

On line 61, there is a reference to a `status` column:

```
    let query = knex
      .select('submission_upload_id', 'submission_id', 'upload_id', 'ua.status')
      .from('submission_upload')
      .join('upload_artifact ua', 'ua.upload_id', 'submission_upload.upload_id')
      .where('submission_id', submissionId);
```

`upload_artifact` doesn't have a `status` column:

```
CREATE TABLE upload_artifact (
  upload_artifact_id uuid DEFAULT public.gen_random_uuid(),
  upload_id uuid NOT NULL,
  artifact_id uuid NOT NULL,
  role upload_artifact_role NOT NULL,
  upload_archive_id uuid,
  ...
);
```

dylanrogowsky-oxd committed:
Remove reference to status column
